README.md
# Retreon

Retreon is a web application to watch over RetroAchievements.org's api in real time and display its updates in a fancy way.

This is a fork from original project created by @spaceglace. The original development was stopped 2 years ago, so i decided to work on it a little more as an opportunity to learn about it's framework (VueJs) and Golang.



CHANGELOG.md
# Change Log
## 2023-03-13
- [x] Polished and fixed some UI behaviour.
- [x] Added visual glowing borders for rare achievements.
- [x] Followed some retroachievements api changes.
- [x] Changed the way sorting achievements works.

## 2022-07-20
- [x] Added a new `Achievements` tab, where you can see all achievements, rarity and dificulty.
- [x] Added option for audio feedback on achievement popup.
- [x] Fixed a bug where achievement popup won't stay on screen.
- [x] Added tooltip options for title and description whenever only icons appears.
- [x] Fixed achievement order to match retroachievements site.
- [x] Added `MASTERED` indicator on focused achievement widget when it's the case.
- [x] Fixed other minor bugs and QOL improvements.


server/main.go
package main

import (
	"net/http"
	"os"
	"path/filepath"
	"retro/config"
	"time"

	"github.com/gorilla/handlers"
	"github.com/gorilla/mux"

	"go.uber.org/zap"
)

var logger *zap.Logger

func serve(content string, handle http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Add("Content-Type", content)
		handle.ServeHTTP(w, r)
	})
}

func main() {
	// set up logger
	logger, _ = zap.NewDevelopment()

	config.Initialize(logger)

	router := mux.NewRouter()

	// set up api endpoints
	router.HandleFunc("/api/game", getGameInformation)
	router.HandleFunc("/api/game/order", setGameOrder)
	router.HandleFunc("/api/metadata", getMetadata)
	router.HandleFunc("/api/metadata/mode", setGameMode)
	router.HandleFunc("/api/metadata/refresh", setRefresh)
	router.HandleFunc("/api/layout/list", getLayoutList)
	router.HandleFunc("/api/layout/add", addLayout)
	router.HandleFunc("/api/layout/remove", removeLayout)
	router.HandleFunc("/api/layout/update", updateLayout)
	router.HandleFunc("/api/layout/active", getActiveLayout)
	router.HandleFunc("/api/layout/active/set", setActiveLayout)
	router.HandleFunc("/api/profile/list", getProfiles)
	router.HandleFunc("/api/profile/add", addProfile)
	router.HandleFunc("/api/profile/remove", removeLayout)
	router.HandleFunc("/api/profile/active", getActiveProfile)
	router.HandleFunc("/api/profile/active/set", setActiveProfile)

	// calculate our working directory
	dir, err := filepath.Abs(filepath.Dir(os.Args[0]))
	if err != nil {
		logger.Fatal("Could not calculate directory",
			zap.Error(err),
		)
	}
	// static directory to serve frontend
	fs := http.FileServer(http.Dir(dir + "/ui"))
	//http.Handle("/", fs)

	router.PathPrefix("/css").Handler(serve(
		"text/css; charset=utf-8",
		http.FileServer(http.Dir(dir+"/ui")),
	))
	router.PathPrefix("/js").Handler(serve(
		"text/javascript; charset=utf-8",
		http.FileServer(http.Dir(dir+"/ui")),
	))
	router.PathPrefix("/").Handler(fs)

	// wait a second then open frontend
	go func() {
		time.Sleep(1 * time.Second)
		logger.Info("Starting server at http://localhost:3000/")
	}()

	// start the server itself
	srv := &http.Server{
		ReadTimeout:  5 * time.Second,
		WriteTimeout: 5 * time.Second,
		Addr:         ":3000",
		Handler: handlers.CORS(
			handlers.AllowedHeaders([]string{"X-Requested-With", "Content-Type", "Authorization"}),
			handlers.AllowedMethods([]string{"GET", "POST", "PUT", "HEAD", "OPTIONS"}),
			handlers.AllowedOrigins([]string{"*"}),
		)(router),
	}
	err = srv.ListenAndServe()
	if err != nil {
		logger.Fatal("Closed http server",
			zap.Error(err),
		)
	}
}


server/config/layout.go
package config

import "fmt"

func GetLayouts() map[string]string {
	if len(settings.Layouts) == 0 {
		AddLayout(
			"default",
			`{"name":"default","width":300,"height":600,"background":"#1E1E1E","auto":false,"interval":30,"widgets":[]}`,
		)
	}
	return settings.Layouts
}

func AddLayout(name string, serialized string) error {
	// check if the name is already in use
	if _, ok := settings.Layouts[name]; ok {
		return fmt.Errorf("name '%s' is already in use", name)
	}

	settings.Layouts[name] = serialized
	return save()
}

func RemoveLayout(name string) error {
	delete(settings.Layouts, name)
	return save()
}

func UpdateLayout(name string, serialized string) error {
	settings.Layouts[name] = serialized
	return save()
}

func GetActiveLayout() string {
	if settings.Layout == "" {
		return "default"
	}
	return settings.Layout
}

func SetActiveLayout(name string) error {
	// ensure layout exists
	if _, ok := settings.Layouts[name]; !ok {
		return fmt.Errorf("layout '%s' does not exist", name)
	}
	settings.Layout = name
	return save()
}


frontend/src/components/widgets/index.js
export default {
  'widget-text': {
    name: 'widget-text',
    title: 'Static Text',
    settings: [
      {
        name: 'text',
        title: 'Text',
        description: 'Static Text',
        type: 'string',
        default: '',
      },
    ],
  },
  'widget-title': {
    name: 'widget-title',
    title: 'Title',
    settings: [
      {
        name: 'height',
        title: 'Height',
        type: 'slider',
        default: 200,
        min: 100,
        max: 800,
      },
      {
        name: 'icon',
        title: 'Image',
        type: 'selection',
        default: 'icon',
        options: [
          { text: 'Icon', value: 'icon' },
          { text: 'Title', value: 'titlescreen' },
          { text: 'In-game', value: 'ingame' },
          { text: 'Box Art', value: 'boxart' },
        ],
      },
      {
        name: 'beautify',
        title: 'Format Title',
        type: 'boolean',
        default: true,
      },
    ],
  },
  'widget-progress': {
    name: 'widget-progress',
    title: 'Progress Bars',
    settings: [
      {
        name: 'showCount',
        title: 'Show Count',
        type: 'boolean',
        default: true,
      },
      {
        name: 'showPoints',
        title: 'Show Points',
        type: 'boolean',
        default: true,
      },
      {
        name: 'showRatio',
        title: 'Show Retro Ratio',
        type: 'boolean',
        default: true,
      },
    ],
  },
  'widget-unlocked': {
    name: 'widget-unlocked',
    title: 'Recent Unlocks',
    settings: [
      {
        name: 'showTooltips',
        title: 'Show Tooltips',
        type: 'boolean',
        default: true,
      },
      {
        name: 'count',
        title: 'Count',
        type: 'slider',
        default: 8,
        min: 0,
        max: 50,
      },
      {
        name: 'iconHeight',
        title: 'Icon Height',
        type: 'slider',
        default: 64,
        min: 12,
        max: 128,
      },
    ],
  },
  'widget-locked': {
    name: 'widget-locked',
    title: 'Next Locked Achievements',
    settings: [
      {
        name: 'showTooltips',
        title: 'Show Tooltips',
        type: 'boolean',
        default: true,
      },
      {
        name: 'count',
        title: 'Count',
        type: 'slider',
        default: 8,
        min: 0,
        max: 50,
      },
      {
        name: 'iconHeight',
        title: 'Icon Height',
        type: 'slider',
        default: 64,
        min: 12,
        max: 128,
      },
    ],
  },
  'widget-focused': {
    name: 'widget-focused',
    title: 'Focused Achievement',
    settings: [
      {
        name: 'marquee',
        title: 'Scroll Description',
        type: 'boolean',
        default: true,
      },
      {
        name: 'softPercent',
        title: 'Show Softcore Percent',
        type: 'boolean',
        default: true,
      },
      {
        name: 'hardPercent',
        title: 'Show Hardcore Percent',
        type: 'boolean',
        default: true,
      },
    ],
  },
  'widget-achievement-overlay': {
    name: 'widget-achievement-overlay',
    title: 'Achievement Overlay',
    settings: [
      {
        name: 'testAchieve',
        title: 'Show Test Overlay',
        type: 'boolean',
        default: false,
      },
      {
        name: 'playUnlockedSound',
        title: 'Play Sound',
        type: 'boolean',
        default: true,
      },
      {
        name: 'shakeType',
        title: 'Animation',
        type: 'selection',
        default: 'shake-chunk',
        options: [
          { text: 'None', value: '' },
          { text: 'Basic', value: 'shake' },
          { text: 'Slow', value: 'shake-slow' },
          { text: 'Little', value: 'shake-little' },
          { text: 'Rotate', value: 'shake-rotate' },
          { text: 'Opacity', value: 'shake-opacity' },
          { text: 'Chunk', value: 'shake-chunk' },
        ],
      },
    ],
  },
  'widget-overlay': {
    name: 'widget-overlay',
    title: 'Static Overlay',
    settings: [
      {
        name: 'text',
        title: 'Text',
        type: 'textarea',
        default: '',
      },
    ],
  },
  'widget-circle': {
    name: 'widget-circle',
    title: 'Circle',
    settings: [
      {
        name: 'height',
        title: 'Height',
        type: 'slider',
        default: 200,
        min: 100,
        max: 800,
      },
    ],
  },
  'widget-spacer': {
    name: 'widget-spacer',
    title: 'Spacer',
    settings: [
      {
        name: 'line',
        title: 'Show Divider',
        type: 'boolean',
        default: true,
      },
      {
        name: 'top',
        title: 'Space Above',
        type: 'slider',
        default: 20,
        min: 0,
        max: 100,
      },
      {
        name: 'bottom',
        title: 'Space Below',
        type: 'slider',
        default: 20,
        min: 0,
        max: 100,
      },
    ],
  },
  'widget-image': {
    name: 'widget-image',
    title: 'Image',
    settings: [
      {
        name: 'url',
        title: 'URL',
        type: 'string',
        default: '',
      },
      {
        name: 'height',
        title: 'Height',
        type: 'slider',
        default: 100,
        min: 20,
        max: 500,
      },
    ],
  },
};


server/retroachievements/api.go
package retroachievements

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"net/http"
	"retro/config"
	"sort"
	"strconv"

	"go.uber.org/zap"
)

const (
	baseURL = "https://retroachievements.org/API"
)

type Game struct {
	ID           string                 `json:"id"`
	Title        string                 `json:"title"`
	Icon         string                 `json:"icon"`
	TitleScreen  string                 `json:"titlescreen"`
	InGame       string                 `json:"ingame"`
	BoxArt       string                 `json:"boxart"`
	Console      string                 `json:"console"`
	Softcore     string                 `json:"softcore"`
	Hardcore     string                 `json:"hardcore"`
	Achievements map[string]Achievement `json:"achievements"`
	Order        []string               `json:"order"`
	DefaultOrder []string               `json:"defaultOrder"`
}

type Achievement struct {
	ID                 string `json:"ID"`
	Badge              string `json:"BadgeName"`
	Title              string `json:"Title"`
	Description        string `json:"Description"`
	Softcore           string `json:"NumAwarded"`
	Hardcore           string `json:"NumAwardedHardcore"`
	Points             string `json:"Points"`
	TrueRatio          string `json:"TrueRatio"`
	DisplayOrder       string `json:"DisplayOrder"`
	DateEarned         string `json:"DateEarned,omitempty"`
	DateEarnedHardcore string `json:"DateEarnedHardcore,omitempty"`
}

func GetLastGame(l *zap.Logger, username, apikey string) (string, error) {
	url := fmt.Sprintf(
		"%s/API_GetUserRecentlyPlayedGames.php?z=%s&y=%s&u=%s&c=1",
		baseURL,
		username,
		apikey,
		username,
	)
	response, err := http.Get(url)
	if code := response.StatusCode; err != nil || code < 200 || code >= 300 {
		l.Error("Error getting recently played games",
			zap.String("username", username),
			zap.Int("code", code),
			zap.Error(err),
		)
		return "", err
	}

	responseBody, err := ioutil.ReadAll(response.Body)
	if err != nil {
		l.Error("Error reading response of recently played games",
			zap.String("username", username),
			zap.Error(err),
		)
		return "", err
	}

	var container []struct {
		GameID int `json:"GameID"`
	}
	if err := json.Unmarshal(responseBody, &container); err != nil {
		l.Error("Error parsing payload of recently played games",
			zap.String("username", username),
			zap.Error(err),
		)
		return "", err
	}

	return strconv.Itoa(container[0].GameID), nil
}

func GetGameInformation(l *zap.Logger, username, apikey string) (Game, error) {
	gameid, err := GetLastGame(l, username, apikey)
	if err != nil {
		l.Error("Error getting last game for user",
			zap.String("username", username),
			zap.Error(err),
		)
		return Game{}, err
	}

	url := fmt.Sprintf(
		"%s/API_GetGameInfoAndUserProgress.php?z=%s&y=%s&u=%s&g=%s",
		baseURL,
		username,
		apikey,
		username,
		gameid,
	)
	response, err := http.Get(url)
	if code := response.StatusCode; err != nil || code < 200 || code >= 300 {
		l.Error("Error getting game information for user",
			zap.String("username", username),
			zap.Int("code", code),
			zap.Error(err),
		)
		return Game{}, err
	}

	responseBody, err := ioutil.ReadAll(response.Body)
	if err != nil {
		l.Error("Error reading game information for user",
			zap.String("username", username),
			zap.Error(err),
		)
		return Game{}, err
	}

	var container struct {
		Title        string                 `json:"Title"`
		Console      string                 `json:"ConsoleName"`
		Icon         string                 `json:"ImageIcon"`
		Titlescreen  string                 `json:"ImageTitle"`
		InGame       string                 `json:"ImageIngame"`
		BoxArt       string                 `json:"ImageBoxArt"`
		Softcore     string                 `json:"NumDistinctPlayersCasual"`
		Hardcore     string                 `json:"NumDistinctPlayersHardcore"`
		Achievements map[string]Achievement `json:"Achievements,omitempty"`
	}
	if err := json.Unmarshal(responseBody, &container); err != nil {
		l.Error("Error parsing game information for user",
			zap.String("username", username),
			zap.Error(err),
		)
		return Game{}, err
	}
	result := Game{
		ID:           gameid,
		Title:        container.Title,
		Console:      container.Console,
		Icon:         container.Icon,
		TitleScreen:  container.Titlescreen,
		InGame:       container.InGame,
		BoxArt:       container.BoxArt,
		Softcore:     container.Softcore,
		Hardcore:     container.Hardcore,
		Achievements: container.Achievements,
	}

	//Set default order
	type order struct {
		ID           int
		DisplayOrder int
	}

	var defaultOrder []order

	for _, achievement := range result.Achievements {
		thisID, _ := strconv.Atoi(achievement.ID)
		thisDisplayOrder, _ := strconv.Atoi(achievement.DisplayOrder)

		thisOrder := order{
			ID:           thisID,
			DisplayOrder: thisDisplayOrder,
		}

		defaultOrder = append(defaultOrder, thisOrder)
	}

	//first sort by ID
	sort.SliceStable(defaultOrder, func(i, j int) bool {
		return defaultOrder[i].ID < defaultOrder[j].ID
	})
	//then sort by display order
	sort.SliceStable(defaultOrder, func(i, j int) bool {
		return defaultOrder[i].DisplayOrder < defaultOrder[j].DisplayOrder
	})
	//save
	for _, defOrder := range defaultOrder {
		result.DefaultOrder = append(result.DefaultOrder, strconv.Itoa(defOrder.ID))
	}

	// look for a saved game order for achievements
	if order, ok := config.GetSettings().Orders[result.ID]; ok {
		result.Order = order
	} else {
		// no saved order, sort in display order
		result.Order = result.DefaultOrder
	}
	return result, nil
}
